pkg/containers-publisher/containers-publisher.go

```diff
@@ -0,0 +1,135 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package containerspublisher
+
+import (
+	"fmt"
+
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
+
+	containercollection "github.com/inspektor-gadget/inspektor-gadget/pkg/container-collection"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+)
+
+type ContainersPublisher struct {
+	collection *containercollection.ContainerCollection
+
+	containersDs              datasource.DataSource
+	eventTypeField            datasource.FieldAccessor
+	cgroupIDField             datasource.FieldAccessor
+	mountNsIDField            datasource.FieldAccessor
+	nameField                 datasource.FieldAccessor
+	containersSubscriptionKey string
+}
+
+func NewContainersPublisher(gadgetCtx operators.GadgetContext, collection *containercollection.ContainerCollection) (*ContainersPublisher, error) {
+	publisher := &ContainersPublisher{}
+	var err error
+
+	publisher.containersDs, err = gadgetCtx.RegisterDataSource(datasource.TypeSingle, "containers")
+	if err != nil {
+		return nil, fmt.Errorf("creating datasource: %w", err)
+	}
+	publisher.containersDs.AddAnnotation("cli.default-output-mode", "none")
+
+	publisher.eventTypeField, err = publisher.containersDs.AddField("event_type", api.Kind_String)
+	if err != nil {
+		return nil, fmt.Errorf("adding field event_type: %w", err)
+	}
+
+	publisher.cgroupIDField, err = publisher.containersDs.AddField("cgroup_id", api.Kind_Uint64)
+	if err != nil {
+		return nil, fmt.Errorf("adding field cgroup_id: %w", err)
+	}
+
+	publisher.mountNsIDField, err = publisher.containersDs.AddField("mntns_id", api.Kind_Uint64)
+	if err != nil {
+		return nil, fmt.Errorf("adding field mntns_id: %w", err)
+	}
+
+	publisher.nameField, err = publisher.containersDs.AddField("name", api.Kind_String)
+	if err != nil {
+		return nil, fmt.Errorf("adding field name: %w", err)
+	}
+
+	publisher.collection = collection
+
+	return publisher, nil
+}
+
+func (publisher *ContainersPublisher) emitContainersDatasourceEvent(eventType containercollection.EventType, container *containercollection.Container, k8s bool) error {
+	ev, err := publisher.containersDs.NewPacketSingle()
+	if err != nil {
+		return fmt.Errorf("creating new containers datasource packet: %w", err)
+	}
+
+	publisher.eventTypeField.PutString(ev, eventType.String())
+	publisher.cgroupIDField.PutUint64(ev, container.CgroupID)
+	publisher.mountNsIDField.PutUint64(ev, container.Mntns)
+	if k8s {
+		publisher.nameField.PutString(ev, container.K8s.ContainerName)
+	} else {
+		publisher.nameField.PutString(ev, container.Runtime.ContainerName)
+	}
```

Comment:
It's fine for now, but ideally both fields should be sent and the consumer should choose which one to consume.